File: code/backend/file_parser.py
# file_parser.py
import pandas as pd
import io
from typing import List
from fastapi import UploadFile
import uuid

from models import PatientRecord


class ParsedFileResult:
    def __init__(self, filename: str, size: int, records: List[PatientRecord], errors: List[str]):
        self.filename = filename
        self.size = size
        self.records = records
        self.errors = errors


async def parse_uploaded_files(files: List[UploadFile]) -> List[ParsedFileResult]:
    """Parse multiple uploaded files (CSV, XLSX, JSON)."""
    results = []
    for file in files:
        content = await file.read()
        size = len(content)
        ext = file.filename.split('.')[-1].lower()
        
        if ext == 'csv':
            records, errors = parse_csv(content, file.filename)
        elif ext in ['xlsx', 'xls']:
            records, errors = parse_excel(content, file.filename)
        elif ext == 'json':
            records, errors = parse_json(content, file.filename)
        else:
            records, errors = [], [f"Unsupported file type: .{ext}"]
        
        results.append(ParsedFileResult(file.filename, size, records, errors))
    
    return results


def parse_csv(content: bytes, filename: str) -> tuple:
    """Parse CSV file into PatientRecord list."""
    try:
        df = pd.read_csv(io.BytesIO(content))
        return df_to_records(df, filename)
    except Exception as e:
        return [], [f"CSV parsing error: {str(e)}"]


def parse_excel(content: bytes, filename: str) -> tuple:
    """Parse Excel file into PatientRecord list."""
    try:
        df = pd.read_excel(io.BytesIO(content), engine='openpyxl')
        return df_to_records(df, filename)
    except Exception as e:
        return [], [f"Excel parsing error: {str(e)}"]


def parse_json(content: bytes, filename: str) -> tuple:
    """Parse JSON file into PatientRecord list."""
    try:
        import json
        data = json.loads(content.decode('utf-8'))
        if isinstance(data, list):
            df = pd.DataFrame(data)
        elif isinstance(data, dict) and 'records' in data:
            df = pd.DataFrame(data['records'])
        else:
            return [], ["JSON must be an array of records or have 'records' field"]
        return df_to_records(df, filename)
    except Exception as e:
        return [], [f"JSON parsing error: {str(e)}"]


def df_to_records(df: pd.DataFrame, filename: str) -> tuple:
    """Convert DataFrame to PatientRecord list with validation."""
    errors = []
    records = []
    
    # Normalize column names
    df.columns = [str(c).lower().strip().replace(' ', '_') for c in df.columns]
    
    # Required column mappings
    required = ['arrival_time', 'service_start', 'service_end']
    missing = [r for r in required if r not in df.columns]
    if missing:
        errors.append(f"Missing columns: {', '.join(missing)}")
        return [], errors
    
    # Fill optional columns with defaults
    if 'patient_id' not in df.columns:
        df['patient_id'] = [f"IMP-{i:05d}" for i in range(len(df))]
    if 'priority' not in df.columns:
        df['priority'] = 3
    if 'department' not in df.columns:
        df['department'] = 'General'
    if 'server_id' not in df.columns:
        df['server_id'] = 1
    
    for idx, row in df.iterrows():
        try:
            arrival = time_to_minutes(row['arrival_time'])
            service_start = time_to_minutes(row['service_start'])
            service_end = time_to_minutes(row['service_end'])
            
            if service_end <= service_start:
                errors.append(f"Row {idx+2}: Invalid service times (end <= start)")
                continue
            
            record = PatientRecord(
                id=str(row['patient_id']),
                arrivalTime=float(arrival),
                serviceStart=float(service_start),
                serviceEnd=float(service_end),
                serverId=int(row.get('server_id', 1)),
                waitTime=max(0.0, service_start - arrival),
                serviceTime=service_end - service_start,
                priority=int(row.get('priority', 3)),
                department=str(row.get('department', 'General'))
            )
            records.append(record)
        except Exception as e:
            errors.append(f"Row {idx+2}: {str(e)}")
            if len(errors) > 20:
                errors.append("... more errors truncated")
                break
    
    return records, errors[:20]


def time_to_minutes(time_val) -> float:
    """Convert various time formats to minutes since midnight."""
    if isinstance(time_val, (int, float)):
        return float(time_val)
    
    time_str = str(time_val).strip()
    # Handle HH:MM:SS or HH:MM
    if ':' in time_str:
        parts = time_str.split(':')
        hours = int(parts[0])
        minutes = int(parts[1]) if len(parts) > 1 else 0
        seconds = int(parts[2]) if len(parts) > 2 else 0
        return hours * 60 + minutes + seconds / 60.0
    
    # Try as numeric string
    try:
        return float(time_str)
    except:
        return 0.0


def anonymize_records(records: List[PatientRecord]) -> List[PatientRecord]:
    """Remove PII by replacing IDs with anonymized tokens."""
    anonymized = []
    for i, rec in enumerate(records):
        anonymized.append(PatientRecord(
            id=f"ANON-{i+1:06d}",
            arrivalTime=rec.arrivalTime,
            serviceStart=rec.serviceStart,
            serviceEnd=rec.serviceEnd,
            serverId=rec.serverId,
            waitTime=rec.waitTime,
            serviceTime=rec.serviceTime,
            priority=rec.priority,
            department=rec.department
        ))
    return anonymized

File: code/client/src/utils/fileParser.ts
/**
 * fileParser.ts
 * Parses CSV / JSON hospital queue data files into PatientRecord arrays.
 */
import type { PatientRecord } from '../types/results';

export interface ParseResult {
  records: PatientRecord[];
  errors: string[];
  rowCount: number;
  skippedRows: number;
}

export interface RawRow {
  patient_id?: string;
  arrival_time?: string;
  service_start?: string;
  service_end?: string;
  department?: string;
  priority?: string;
  server_id?: string;
  [key: string]: string | undefined;
}

// ─── CSV ─────────────────────────────────────────────────────────────────────
export function parseCSV(text: string): RawRow[] {
  const lines = text.trim().split('\n').filter(Boolean);
  if (lines.length < 2) return [];
  const headers = lines[0].split(',').map(h => h.trim().toLowerCase().replace(/\s+/g, '_'));
  return lines.slice(1).map(line => {
    const values = line.split(',').map(v => v.trim());
    const row: RawRow = {};
    headers.forEach((h, i) => { row[h] = values[i] ?? ''; });
    return row;
  });
}

// ─── JSON ────────────────────────────────────────────────────────────────────
export function parseJSON(text: string): RawRow[] {
  try {
    const data = JSON.parse(text);
    return Array.isArray(data) ? data : data.records ?? data.patients ?? [];
  } catch {
    return [];
  }
}

// ─── Validation & mapping ────────────────────────────────────────────────────
function toMinutes(timeStr: string): number {
  if (!timeStr) return 0;
  const parts = timeStr.includes('T')
    ? timeStr.split('T')[1].split(':')
    : timeStr.split(':');
  const h = parseInt(parts[0], 10) || 0;
  const m = parseInt(parts[1], 10) || 0;
  return h * 60 + m;
}

export function mapRowToPatient(row: RawRow, index: number): PatientRecord | null {
  const id = row.patient_id ?? row.id ?? `P-${String(index + 1).padStart(4, '0')}`;
  const arrival = toMinutes(row.arrival_time ?? '');
  const start   = toMinutes(row.service_start ?? row.start_time ?? '');
  const end     = toMinutes(row.service_end ?? row.end_time ?? '');
  if (end <= start) return null;
  return {
    id,
    arrivalTime:  arrival,
    serviceStart: start,
    serviceEnd:   end,
    serverId:     parseInt(row.server_id ?? row.server ?? '1', 10),
    waitTime:     Math.max(0, start - arrival),
    serviceTime:  end - start,
    priority:     parseInt(row.priority ?? '1', 10),
    department:   row.department ?? row.unit ?? 'General',
  };
}

// ─── Main entry ──────────────────────────────────────────────────────────────
export async function parseFile(file: File): Promise<ParseResult> {
  const text = await file.text();
  const errors: string[] = [];
  let rawRows: RawRow[] = [];

  const ext = file.name.split('.').pop()?.toLowerCase();
  if (ext === 'csv') {
    rawRows = parseCSV(text);
  } else if (ext === 'json') {
    rawRows = parseJSON(text);
  } else if (ext === 'xlsx' || ext === 'xls') {
    // For Excel, we would need a library; but your backend handles Excel.
    // If you want client-side Excel parsing, add xlsx library.
    errors.push(`Client-side Excel parsing not supported. Use backend or convert to CSV.`);
    return { records: [], errors, rowCount: 0, skippedRows: 0 };
  } else {
    errors.push(`Unsupported file type: .${ext}. Use CSV or JSON.`);
    return { records: [], errors, rowCount: 0, skippedRows: 0 };
  }

  const records: PatientRecord[] = [];
  let skipped = 0;
  rawRows.forEach((row, i) => {
    const patient = mapRowToPatient(row, i);
    if (patient) {
      records.push(patient);
    } else {
      skipped++;
      if (skipped <= 5) errors.push(`Row ${i + 2}: Invalid or missing time data — skipped.`);
    }
  });
  if (skipped > 5) errors.push(`…and ${skipped - 5} more rows skipped.`);

  return { records, errors, rowCount: rawRows.length, skippedRows: skipped };
}

// ─── Anonymisation ───────────────────────────────────────────────────────────
export function anonymiseRecords(records: PatientRecord[]): PatientRecord[] {
  return records.map((r, i) => ({
    ...r,
    id: `ANON-${String(i + 1).padStart(5, '0')}`,
  }));
}